File: app/src/main/java/com/boardgamegeek/tasks/sync/SyncTask.java
package com.boardgamegeek.tasks.sync;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.AsyncTask;

import com.boardgamegeek.R;
import com.boardgamegeek.extensions.IntUtils;
import com.boardgamegeek.extensions.NetworkUtils;
import com.boardgamegeek.io.Adapter;
import com.boardgamegeek.io.BggService;
import com.boardgamegeek.util.RemoteConfig;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.lifecycle.MutableLiveData;
import retrofit2.Call;
import retrofit2.Response;
import timber.log.Timber;

public abstract class SyncTask<T> extends AsyncTask<Void, Void, String> {
	@SuppressLint("StaticFieldLeak") @NonNull protected final Context context;
	protected final long startTime;
	protected BggService bggService;
	private Call<T> call;
	private int page = 1;

	public final MutableLiveData<String> errorMessageLiveData = new MutableLiveData<>();

	SyncTask(@NonNull Context context) {
		this.context = context.getApplicationContext();
		startTime = System.currentTimeMillis();
	}

	@Override
	protected String doInBackground(Void... params) {
		bggService = createService();
		if (!isRequestParamsValid())
			return context.getString(R.string.msg_update_invalid_request, context.getString(getTypeDescriptionResId()));
		if (NetworkUtils.isOffline(context)) return context.getString(R.string.msg_offline);
		if (!RemoteConfig.getBoolean(RemoteConfig.KEY_SYNC_ENABLED))
			return context.getString(R.string.msg_sync_remotely_disabled);
		try {
			boolean hasMorePages;
			page = 0;
			do {
				page++;
				call = createCall();
				Response<T> response = call.execute();
				if (response.isSuccessful()) {
					if (isResponseBodyValid(response.body())) {
						persistResponse(response.body());
					} else {
						return context.getString(R.string.msg_update_invalid_response,
							context.getString(getTypeDescriptionResId())); // TODO: 3/26/17 include key
					}
				} else {
					Timber.w("Received response %s while syncing %s.", response.code(), context.getString(getTypeDescriptionResId()));
					return IntUtils.asHttpErrorMessage(response.code(), context);
				}
				if (isCancelled()) break;
				hasMorePages = hasMorePages(response.body());
			} while (hasMorePages);
			if (!isCancelled()) finishSync();
		} catch (Exception e) {
			Timber.w(e, "Exception fetching %1$s: %2$s", context.getString(getTypeDescriptionResId()), e.getMessage());
			return e.getLocalizedMessage();
		}
		return "";
	}

	@Override
	protected void onPostExecute(String errorMessage) {
		Timber.w(errorMessage);
		errorMessageLiveData.postValue(errorMessage);
	}

	@Override
	protected void onCancelled() {
		if (call != null) call.cancel();
	}

	protected int getCurrentPage() {
		return page;
	}

	@StringRes
	protected abstract int getTypeDescriptionResId();

	protected abstract Call<T> createCall();

	protected BggService createService() {
		return Adapter.createForXml();
	}

	protected abstract boolean isRequestParamsValid();

	protected boolean isResponseBodyValid(T body) {
		return body != null;
	}

	protected abstract void persistResponse(T body);

	protected boolean hasMorePages(T body) {
		return false;
	}

	protected void finishSync() {
	}
}

File: app/src/main/java/com/boardgamegeek/tasks/sync/SyncTask.kt
package com.boardgamegeek.tasks.sync

import android.annotation.SuppressLint
import android.content.Context
import android.os.AsyncTask
import androidx.annotation.StringRes
import androidx.lifecycle.MutableLiveData
import com.boardgamegeek.R
import com.boardgamegeek.extensions.asHttpErrorMessage
import com.boardgamegeek.extensions.isOffline
import com.boardgamegeek.io.Adapter
import com.boardgamegeek.io.BggService
import com.boardgamegeek.util.RemoteConfig
import com.boardgamegeek.util.RemoteConfig.Companion.getBoolean
import retrofit2.Call
import timber.log.Timber

abstract class SyncTask<T> internal constructor(context: Context) : AsyncTask<Void, Void, String>() {
    @SuppressLint("StaticFieldLeak")
    protected val context: Context = context.applicationContext

    protected val startTime: Long = System.currentTimeMillis()

    protected var bggService: BggService? = null

    private var call: Call<T>? = null

    protected var currentPage = 1
        private set

    private val errorMessageLiveData = MutableLiveData<String>()

    override fun doInBackground(vararg params: Void): String {
        bggService = createService()
        if (!isRequestParamsValid) return context.getString(R.string.msg_update_invalid_request, context.getString(typeDescriptionResId))
        if (context.isOffline()) return context.getString(R.string.msg_offline)
        if (!getBoolean(RemoteConfig.KEY_SYNC_ENABLED)) return context.getString(R.string.msg_sync_remotely_disabled)
        try {
            var shouldContinue = false
            currentPage = 0
            do {
                currentPage++
                call = createCall()
                val response = call?.execute()
                if (response == null) {
                    Timber.w("Null response while syncing %s.", context.getString(typeDescriptionResId))
                } else {
                    if (response.isSuccessful) {
                        if (isResponseBodyValid(response.body())) {
                            persistResponse(response.body())
                        } else {
                            return context.getString(R.string.msg_update_invalid_response,
                                    context.getString(typeDescriptionResId))
                        }
                    } else {
                        Timber.w("Received response %s while syncing %s.", response.code(), context.getString(typeDescriptionResId))
                        return response.code().asHttpErrorMessage(context)
                    }
                    shouldContinue = hasMorePages(response.body())
                }
                if (isCancelled) break
            } while (shouldContinue)
            if (!isCancelled) finishSync()
        } catch (e: Exception) {
            Timber.w(e, "Exception fetching %1\$s: %2\$s", context.getString(typeDescriptionResId), e.message)
            return e.localizedMessage ?: e.message ?: e.toString()
        }
        return ""
    }

    override fun onPostExecute(errorMessage: String) {
        Timber.w(errorMessage)
        errorMessageLiveData.postValue(errorMessage)
    }

    override fun onCancelled() {
        call?.cancel()
    }

    @get:StringRes
    protected abstract val typeDescriptionResId: Int

    protected abstract fun createCall(): Call<T>?

    protected open fun createService(): BggService {
        return Adapter.createForXml()
    }

    protected abstract val isRequestParamsValid: Boolean

    open fun isResponseBodyValid(body: T?): Boolean {
        return body != null
    }

    protected abstract fun persistResponse(body: T?)

    protected open fun hasMorePages(body: T?): Boolean {
        return false
    }

    protected open fun finishSync() {}
}